python/pyarrow/array.pxi

```diff
@@ -1510,6 +1526,37 @@ cdef class DictionaryArray(Array):
         result.init(c_result)
         return result
 
+cdef class StructArray(Array):
+    @staticmethod
+    def from_arrays(field_names, arrays):
+        cdef:
+            Array array
+            shared_ptr[CArray] c_array
+            vector[shared_ptr[CArray]] c_arrays
+            shared_ptr[CArray] c_result
+
+        if len(arrays) == 0:
+            raise ValueError("arrays list is empty")
+
+        length = len(arrays[0])
+
+        c_arrays.resize(len(arrays))
+        for i, array in enumerate(arrays):
```

Comment:
I believe if you declare `i` as something like `size_t` or `int` then the iteration happens with C for loops rather than with Python's iteration protocol.